src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -18468,6 +18479,400 @@ public void Main()
+    [MaybeNull]public TClass? P3 { get; set; } = null;
+}
+public class CStruct<TStruct> where TStruct : struct
+{
+    [MaybeNull]public TStruct P4 { get; set; }
+    [MaybeNull]public TStruct? P5 { get; set; }
+}";
+            var lib = CreateNullableCompilation(new[] { MaybeNullAttributeDefinition, lib_cs });
+
+            var source = @"
+class C
+{
+    static void M1<T>(T t1)
+    {
+        new COpen<T>().P1.ToString(); // 1
+    }
+    static void M2<T>(T t2) where T : class
+    {
+        new COpen<T>().P1.ToString(); // 2
+        new CClass<T>().P2.ToString(); // 3
+        new CClass<T>().P3.ToString(); // 4
+    }
+    static void M4<T>(T t4) where T : struct
+    {
+        new COpen<T>().P1.ToString();
+        new CStruct<T>().P4.ToString();
+        new COpen<T?>().P1.Value.ToString(); // 5
+        new CStruct<T>().P5.Value.ToString(); // 6
+    }
+}";
+            var expected = new[]
+            {
+                // (6,9): warning CS8602: Dereference of a possibly null reference.
+                //         new COpen<T>().P1.ToString(); // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "new COpen<T>().P1").WithLocation(6, 9),
+                // (10,9): warning CS8602: Dereference of a possibly null reference.
+                //         new COpen<T>().P1.ToString(); // 2
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "new COpen<T>().P1").WithLocation(10, 9),
+                // (11,9): warning CS8602: Dereference of a possibly null reference.
+                //         new CClass<T>().P2.ToString(); // 3
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "new CClass<T>().P2").WithLocation(11, 9),
+                // (12,9): warning CS8602: Dereference of a possibly null reference.
+                //         new CClass<T>().P3.ToString(); // 4
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "new CClass<T>().P3").WithLocation(12, 9),
+                // (18,9): warning CS8629: Nullable value type may be null.
+                //         new COpen<T?>().P1.Value.ToString(); // 5
+                Diagnostic(ErrorCode.WRN_NullableValueTypeMayBeNull, "new COpen<T?>().P1").WithLocation(18, 9),
+                // (19,9): warning CS8629: Nullable value type may be null.
+                //         new CStruct<T>().P5.Value.ToString(); // 6
+                Diagnostic(ErrorCode.WRN_NullableValueTypeMayBeNull, "new CStruct<T>().P5").WithLocation(19, 9)
+            };
+
+            var comp = CreateNullableCompilation(source, references: new[] { lib.EmitToImageReference() });
+            comp.VerifyDiagnostics(expected);
+
+            var comp2 = CreateNullableCompilation(new[] { source, lib_cs, MaybeNullAttributeDefinition });
+            comp2.VerifyDiagnostics(expected);
+
+            CompileAndVerify(comp2, validator: assembly =>
+            {
+                var reader = assembly.GetMetadataReader();
```

Comment:
>var reader = assembly.GetMetadataReader(); [](start = 16, length = 42)

Why not use symbol validator and examine attributes on symbols instead?

Reply:
(answered elsewhere)

---
In reply to: [295457267](https://github.com/dotnet/roslyn/pull/36444#discussion_r295457267) [](ancestors = 295457267)

Reply:
> (answered elsewhere)

I do not believe this is worth added complexity. It just doesn't buy us anything here and complicates understanding and maintenance of the tests. 

---
In reply to: [295493320](https://github.com/dotnet/roslyn/pull/36444#discussion_r295493320) [](ancestors = 295493320,295457267)